File: src/org/genyris/dl/TripleFunctions.java
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.dl;

import org.genyris.core.*;
import org.genyris.exception.GenyrisException;
import org.genyris.interp.AbstractMethod;
import org.genyris.interp.ApplicableFunction;
import org.genyris.interp.Closure;
import org.genyris.interp.Environment;
import org.genyris.interp.Interpreter;
import org.genyris.interp.UnboundException;

public class TripleFunctions extends ApplicableFunction {

    public TripleFunctions(Interpreter interp) {
        super(interp, "triple", true);
    }

    public Symbol toSymbol(Exp x) throws GenyrisException {
        if( x instanceof Bignum ||  x instanceof StrinG ||  x instanceof Symbol ) {
            return _interp.intern(x.toString());
        }
        else {
            throw new GenyrisException( "Cannot make a triple with " + x.toString() + " " + x.getClass().getName());
        }
    }
    public Exp bindAndExecute(Closure proc, Exp[] arguments,
            Environment envForBindOperations) throws GenyrisException {
        checkArguments(arguments, 3);
        Class[] types = { Exp.class, Exp.class, Exp.class };
        checkArgumentTypes(types, arguments);
        return new Triple(toSymbol(arguments[0]), toSymbol(arguments[1]), arguments[2]);
    }

    public static abstract class AbstractTripleMethod extends AbstractMethod {

        public AbstractTripleMethod(Interpreter interp, String name) {
            super(interp, name);
        }

        protected Triple getSelfAsTriple(Environment env) throws GenyrisException {
            getSelf(env);
            if (!(_self instanceof Triple)) {
                throw new GenyrisException(
                        "Non-Triple passed to a Triple method.");
            } else {
                return (Triple) _self;
            }
        }
    }


    public static void bindFunctionsAndMethods(Interpreter interpreter) throws UnboundException, GenyrisException {
        interpreter.bindGlobalProcedureInstance(new TripleFunctions(interpreter));
    }

}

File: src/org/genyris/dl/TripleFunctions.kt
//Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.dl

import org.genyris.core.Bignum
import org.genyris.core.Exp
import org.genyris.core.StrinG
import org.genyris.core.Symbol
import org.genyris.exception.GenyrisException
import org.genyris.interp.*

class TripleFunctions(interp: Interpreter) : ApplicableFunction(interp, "triple", true) {
    @kotlin.Throws(GenyrisException::class)
    override fun toSymbol(x: Exp): Symbol? {
        if (x is Symbol) {
            return _interp.intern(x)
        }
        if (x is Bignum || x is StrinG || x is Symbol) {
            return _interp.intern(x.toString())
        }
        throw GenyrisException("Cannot make a triple with " + x.toString() + " " + x.getClass().getName())
    }

    @kotlin.Throws(GenyrisException::class)
    override fun bindAndExecute(
        proc: Closure?, arguments: Array<Exp?>,
        envForBindOperations: Environment?
    ): Exp {
        checkArguments(arguments, 3)
        val types = arrayOf<Class<*>?>(Exp::class.java, Exp::class.java, Exp::class.java)
        checkArgumentTypes(types, arguments)
        return Triple(toSymbol(arguments[0]!!), toSymbol(arguments[1]!!), arguments[2])
    }

    abstract class AbstractTripleMethod(interp: Interpreter?, name: String?) : AbstractMethod(interp, name) {
        @kotlin.Throws(GenyrisException::class)
        protected fun getSelfAsTriple(env: Environment): Triple {
            getSelf(env)
            if (_self !is Triple) {
                throw GenyrisException(
                    "Non-Triple passed to a Triple method."
                )
            } else {
                return _self as Triple
            }
        }
    }


    companion object {
        @kotlin.Throws(UnboundException::class, GenyrisException::class)
        fun bindFunctionsAndMethods(interpreter: Interpreter) {
            interpreter.bindGlobalProcedureInstance(TripleFunctions(interpreter))
        }
    }
}
